CAMEL-6840 - add grouped throttling feature, XML DSL is not working p…

onderson opened the pull request:
…roperly so it is WIP for and now checking it in as Ignored test case

For now unit test in camel-spring is ignored.
I think i need help with it.
fyi. @davsclaus , @WillemJiang , @oscerd and @aldettinger

aldettinger commented on components/camel-spring/src/test/resources/org/apache/camel/spring/processor/ThrottlerGroupingTest.xml:
```diff
@@ -0,0 +1,69 @@
+<?xml version="1.0" encoding="UTF-8"?>
```
Globally, you have 2 issues. It is not valid against the xml schema, and the test is not the exact translation of the java DSL `ThrottlingGroupingTest` tests. Examples below.

onderson replied:
this is addressed

aldettinger commented on components/camel-spring/src/test/resources/org/apache/camel/spring/processor/ThrottlerGroupingTest.xml:
```diff
@@ -0,0 +1,69 @@
+       http://camel.apache.org/schema/spring http://camel.apache.org/schema/spring/camel-spring.xsd
+    ">
+
+  <bean id="myBean" class="org.apache.camel.spring.processor.SpringThrottlerMethodCallTest"/>
```
No sure it is used/useful

onderson replied:
Yes. True

aldettinger commented on components/camel-spring/src/test/resources/org/apache/camel/spring/processor/ThrottlerGroupingTest.xml:
```diff
@@ -0,0 +1,69 @@
+    </route>
+    
+    <route errorHandlerRef="dlc">
+      <from uri="seda:a"/>
```
The java DSL test states `seda:c` here

aldettinger commented on components/camel-spring/src/test/resources/org/apache/camel/spring/processor/ThrottlerGroupingTest.xml:
```diff
@@ -0,0 +1,69 @@
+        </correlationExpression>
+        <header>max</header>
+        <to uri="log:result"/>
+        <to uri="mock:result"/>
```
The java DSL test states `mock:result2` here

onderson replied:
As xml dsl does not work properly, i did not bother making them the same yet. You are right they should be the same at the end.

aldettinger commented on components/camel-spring/src/test/resources/org/apache/camel/spring/processor/ThrottlerGroupingTest.xml:
```diff
@@ -0,0 +1,69 @@
+    
+    <route errorHandlerRef="dlc">
+      <from uri="seda:a"/>
+      <throttle timePeriodMillis="1000">
```
The java DSL test states 2000ms here

aldettinger commented on components/camel-spring/src/test/resources/org/apache/camel/spring/processor/ThrottlerGroupingTest.xml:
```diff
@@ -0,0 +1,69 @@
+    <errorHandler id="dlc" deadLetterUri="mock:dead" type="DeadLetterChannel"/>
+    <route errorHandlerRef="dlc">
+      <from uri="seda:a"/>
+      <throttle timePeriodMillis="1000">
```
As found by Willem, the current version of the camel-spring schema requires to put the correlationExpression after the last `to` of this route.

onderson replied:
this is addressed with xmltype annotation and propOrder attribute of it.

aldettinger commented on camel-core/src/main/java/org/apache/camel/processor/Throttler.java:
```diff
@@ -61,25 +70,32 @@
     private static final String PROPERTY_EXCHANGE_STATE = "CamelThrottlerExchangeState";
+    // (throttling grouping) defaulted as 1 because there will be only one queue which is similar to implementation
+    // when there is no grouping for throttling
+    private static final Integer NO_CORRELATION_QUEUE_ID = new Integer(1);
```
Does it mean that we mix the default group when a correlationExpression evaluates to 1 ?

onderson replied:
Yes, this is correct. Not sure if that would happen very easily. Could not come up with something brighter. Open to hear ideas.

aldettinger replied:
Maybe keep throttleRate/delayQueue for default group and throttleRatesMap/defaultQueueCache for other groups.
Perhaps, put that in a dedicated class named `ThrottleRates` ?
Another idea could be to use a less common value, e.g. `NO_CORRELATION_QUEUE_ID = new Integer(Integer.MAX_VALUE)`.
Note that having a `correlationExpression` that could evaluate to the default group key would allow dynamic inclusion in the default group.

onderson replied:
nice comments, addressed some of them.
considering the ones(till others) left undone

aldettinger commented on camel-core/src/main/java/org/apache/camel/processor/Throttler.java:
```diff
@@ -192,6 +217,26 @@ public boolean process(final Exchange exchange, final AsyncCallback callback) {
         }
     }
 
+    private DelayQueue<ThrottlePermit> locateDelayQueue(final Integer key) throws InterruptedException, ExecutionException {        
```
Could you please elaborate on why you are using a CompletableFuture here ?

onderson replied:
I could call complete manually, to get the result. That was it.

WillemJiang commented on camel-core/src/main/java/org/apache/camel/model/ThrottleDefinition.java:
```diff
@@ -43,6 +44,8 @@
     // TODO: Camel 3.0 Should not support outputs
 
+    @XmlElement(name = "correlationExpression")
+    private ExpressionSubElementDefinition correlationExpression;
```
This correlationExpression must be put after to element, you had to find a way to setup the right order for it.  I don't think you can do it by just extends the ExpressionNode.

onderson replied:
Not sure if it is a must. We can impose order with xmlorder tag. I did try. It did not help.

davsclaus replied:
Yeah I would also put correlationExpression as 2nd, eg its a new option, and it makes more sense, as its optional.

onderson replied:
this is addressed

WillemJiang commented on camel-core/src/main/java/org/apache/camel/processor/Throttler.java:
```diff
@@ -61,25 +70,32 @@
     private boolean callerRunsWhenRejected = true;
+    private Expression correlationExpression;
+    // below 2 fields added for (throttling grouping)
+    private Map<Integer, DelayQueue<ThrottlePermit>> delayQueueCache;
```
We may introduce memory leak issue, if we keep adding items to the map without clean them.

onderson replied:
It is actually cleared in shutdown. And i dont expect extreme numbers of grouping. It is possible but not so likely

davsclaus replied:
Wont the delay queue empty itself when the time period has elapsed, and then release the exchanges. So even if you have many correlation keys then they would eventually be released. And yeah clean in shutdown is good practice (although Camel ought to do graceful shutdown and wait for those throttled exchanges to complete first).

WillemJiang commented on camel-core/src/main/java/org/apache/camel/processor/Throttler.java:
```diff
@@ -192,6 +217,26 @@ public boolean process(final Exchange exchange, final AsyncCallback callback) {
+        delayQueueCacheExecutorService.submit(() -> {
+            futureDelayQueue.complete(findDelayQueue(key));
+        });
+        DelayQueue<ThrottlePermit> currentQueue = futureDelayQueue.get();   
```
futrueDelayQueue.get() is wait for the future method to complete, it doesn't save us time.

onderson replied:
My intention was not save time, was to fill up the throttle group async

aldettinger replied:
I agree with Willem, I don't see yet the value of doing this async.

onderson replied:
maybe not very big plus. and 
i disagree. In case of async routing, could be useful.

WillemJiang commented on camel-core/src/main/java/org/apache/camel/processor/Throttler.java:
```diff
@@ -235,20 +282,22 @@ protected void enqueuePermit(final ThrottlePermit permit, final Exchange exchang
             throw new IllegalStateException("The maximumRequestsPerPeriod must be a positive number, was: " + newThrottle);
         }
 
         synchronized (this) {
```
As we separate throttleRate by the correlation id, we could use throttleRate as a lock instead of using this throttle object.

onderson replied:
CorrelationExpression may be even better?

aldettinger replied:
My understanding is that `correlationExpression` is always the same. Would the finer grained lock be the result of the evaluation then, so namely `key`?

onderson replied:
Yes i meant the evaluated correlationExpression. I agree

onderson commented:
Addressed review comments and now XML Dsl is working.
Please review again.
thanks for any feedback

onderson commented:
I remember i forgot to address comments for synchronised block. As i will wait for @oscerd and @davsclaus , i will try to handle comments all together.

aldettinger commented on camel-core/src/main/java/org/apache/camel/processor/Throttler.java:
```diff
@@ -111,13 +128,21 @@ public boolean process(final Exchange exchange, final AsyncCallback callback) {
-            calculateAndSetMaxRequestsPerPeriod(exchange);
+            Integer key;
+            if (correlationExpression != null) {
+                key = correlationExpression.evaluate(exchange, Integer.class);
```
If we use `key` as a finer grain lock, I would make a defensive copy because I'm not sure that the reference returned by evaluate could be used as a lock by another thread for another purpose.

onderson replied:
good point

davsclaus commented on camel-core/src/main/java/org/apache/camel/model/ThrottleDefinition.java:
```diff
@@ -55,14 +60,26 @@
         super(maximumRequestsPerPeriod);
     }
 
+    public ThrottleDefinition(Expression correlationExpression, Expression maximumRequestsPerPeriod) {
```
These 2 constructors are using different order of the parameters, eg it would be good to keep it consistent. As correlation expression is a new parameter, I would add it as 2nd.

onderson commented:
rebased and addressed code review comments. Could you guys take a look one again?
Thanks again for the reviews. Yes it is a great involvement. Happy to get feedback..

WillemJiang commented on camel-core/src/main/java/org/apache/camel/processor/Throttler.java:
```diff
@@ -235,63 +283,122 @@ protected void enqueuePermit(final ThrottlePermit permit, final Exchange exchang
+        Integer currentThrottleRate = throttleRate;
+        if (correlationExpression != null) {
+            currentThrottleRate = throttleRatesMap.get(key);
+            lockOnSync = new Integer(currentThrottleRate.intValue());
```
Can we use the throttleRate as the lock here? I don't think create a new Integer can do the job.

aldettinger replied:
Indeed, creating a new Integer each time can't do the job.
Let's use throttleRate as the lock.
My understanding is that all groups will share the same lock, but optimization could come later on.

onderson replied:
@WillemJiang and @aldettinger , guys can you please be more explicit about this subject? I am afraid i did not get what you mean.

WillemJiang replied:
Our suggestion is using the currentThrottleRate as the lock, and if you create a new Integer with the currentThrottleRate intValue, you always get a new Integer instance for the lockOnSync variable. It doesn't  block the thread which try to access same ThrottleRate at the same time.

onderson replied:
I got that but that was the point of @aldettinger's first suggestion to make a defensive copy due to not being sure of the reference returned by evaluate could be used as a lock by another thread for another purpose or not. I think that led me to think like that. I think the correct way of managing synchronised block will be like this
    1- If no grouping applied, throttler itself should be used as lock because all threads will be accessing volatile throttleRate and modify it.
    2- If grouping applied, not the value from throttleRates map should be used as lock but the key would be a better choice.

that was what i already mentioned but reflected in an unintended way to implementation.
https://github.com/apache/camel/pull/2376#discussion_r195896140

aldettinger replied:
@onderson
Sorry for confusion. I think it will be clearer expressing my preferences below:
1) Make a defensive copy of the key only once, and lock on that same object for the whole group life
2) Lock on the value, currentThrottleRate
3) Lock on throttleRate (group would have been mixed)

onderson replied:
That's all right. It was a good catch.
I disagree about defensive copy of the key because those keys (key of throttleRatesMap) are already private and non-accessible variables, and there is no other place those keys used by any other threading mechanism. 
option 2 and 3 and are not good choices already IMHO whereas we can use evaluated value of correlationExpression as the lock for synchronization.

onderson commented:
@WillemJiang , @aldettinger ,  @oscerd , @davsclaus , are you ok with PR?

oscerd commented:
Approved for me.

davsclaus commented:
This PR is a bit hard to grok - especially that code that does that completion future task. For me it looks like you use a thread pool to do some work to find a delay queue, and then do a future get to block and wait for that thread to finish. What is the reason for that?

onderson commented:
I thought like this;
In case of async routing there may be cases where we create new group whose correlationExpression might first hit after long series of exchanges which the same key hit and are to be on hold in their delayQueue. so we need to find delay queue to add new ones while we create a new empty delay queue for the new group hit for the first time. that's why i thought locating delay queues for those frequently hitting exchanges for the group during async routing would be better be async with another executorservice. that was my only reasoning. this may not bring much value if that would not happen and exchanges are evenly distributed to be throttled.

davsclaus commented:
Okay thanks for the details. I think we need to add some code comment about this, or as javadoc to the class etc about implementation details.

Also I would prefer if the thread pool is only created when you use correlationExpression. As then its not needed and just takes up resources. Also we should allow to configure a shared thread pool like some of the other EIPs does. For example splitter, wire-tap etc, with the `ExecutorServiceAwareDefinition`

onderson commented:
OK i will amend PR as requested.
I have a question.
ThrottleDefinition implements ExecutorServiceAwareDefinition already and Throttler has a shared pool, what you means is to use that one to submit those Callables to that pool or not?

onderson commented:
updated anyway, please let me know if it is ok or no.. thanks.
